File: android/app/src/main/java/com/voxtv/firetv/StreamDetect.kt
package com.voxtv.firetv

/**
 * Port of `src/lib/iptv/stream-detect.ts`. The player picks its engine from the
 * kind reported here, so the sniffing rules must stay byte-for-byte identical
 * to the web build.
 */
object StreamDetect {
  const val HLS = "hls"
  const val TS = "ts"
  const val MP4 = "mp4"
  const val MKV = "mkv"
  const val UNKNOWN = "unknown"

  private fun looksLikeTsAligned(bytes: ByteArray, from: Int): Boolean {
    val limit = minOf(bytes.size, from + 2048)
    var i = from
    val scanEnd = minOf(bytes.size, from + 188)
    while (i < scanEnd) {
      if (bytes[i] != 0x47.toByte()) {
        i++
        continue
      }
      val next = i + 188
      val third = i + 376
      if (next >= limit) return true
      if (bytes[next] == 0x47.toByte() && (third >= limit || bytes[third] == 0x47.toByte())) return true
      i++
    }
    return false
  }

  fun sniff(bytes: ByteArray): String {
    if (bytes.isEmpty()) return UNKNOWN
    var offset = 0
    while (offset < bytes.size) {
      val b = bytes[offset]
      if (b == 0x20.toByte() || b == 0x0a.toByte() || b == 0x0d.toByte() || b == 0x09.toByte()) {
        offset++
      } else {
        break
      }
    }
    val remaining = bytes.size - offset
    if (remaining >= 4) {
      val head = String(bytes, offset, 4, Charsets.ISO_8859_1)
      if (head.startsWith("#EXT")) return HLS
      if (head.startsWith("ID3")) {
        val skip = minOf(bytes.size, offset + 10)
        return if (looksLikeTsAligned(bytes, skip)) TS else UNKNOWN
      }
    }
    if (looksLikeTsAligned(bytes, offset)) return TS
    if (remaining >= 8) {
      val brand = String(bytes, offset + 4, 4, Charsets.ISO_8859_1)
      if (brand == "ftyp") return MP4
    }
    if (remaining >= 3 &&
      bytes[offset] == 0x1a.toByte() &&
      bytes[offset + 1] == 0x45.toByte() &&
      bytes[offset + 2] == 0xdf.toByte()
    ) {
      return MKV
    }
    val textLen = minOf(remaining, 24)
    if (textLen > 0) {
      val text = String(bytes, offset, textLen, Charsets.UTF_8).trimStart()
      if (text.startsWith("#EXT")) return HLS
    }
    return UNKNOWN
  }

  fun mimeForKind(kind: String, fallback: String?): String = when (kind) {
    HLS -> "application/vnd.apple.mpegurl"
    TS -> "video/mp2t"
    MP4 -> "video/mp4"
    MKV -> "video/x-matroska"
    else -> if (fallback.isNullOrBlank()) "application/octet-stream" else fallback
  }

  fun looksLikePlaylistUrl(url: String, contentType: String): Boolean {
    if (Regex("mpegurl|m3u8", RegexOption.IGNORE_CASE).containsMatchIn(contentType)) return true
    return Regex("\\.m3u8(\\?|$)", RegexOption.IGNORE_CASE).containsMatchIn(url)
  }
}

File: src/lib/iptv/stream-detect.ts
export type StreamKind = "hls" | "ts" | "mp4" | "mkv" | "unknown";

function looksLikeTsAligned(bytes: Uint8Array): boolean {
  const limit = Math.min(bytes.length, 2048);
  for (let i = 0; i < Math.min(bytes.length, 188); i++) {
    if (bytes[i] !== 0x47) continue;
    const next = i + 188;
    const third = i + 376;
    if (next >= limit) return true;
    if (bytes[next] === 0x47 && (third >= limit || bytes[third] === 0x47)) return true;
  }
  return false;
}

export function sniffStreamBytes(bytes: Uint8Array | Buffer): StreamKind {
  if (!bytes || bytes.length === 0) return "unknown";
  let offset = 0;
  while (
    offset < bytes.length &&
    (bytes[offset] === 0x20 || bytes[offset] === 0x0a || bytes[offset] === 0x0d || bytes[offset] === 0x09)
  ) {
    offset += 1;
  }
  const slice = bytes.subarray(offset);
  if (slice.length >= 4) {
    const head = String.fromCharCode(slice[0]!, slice[1]!, slice[2]!, slice[3]!);
    if (head.startsWith("#EXT")) return "hls";
    if (head.startsWith("ID3")) {
      const rest = slice.subarray(Math.min(slice.length, 10));
      if (looksLikeTsAligned(rest)) return "ts";
      return "unknown";
    }
  }
  if (looksLikeTsAligned(slice)) return "ts";
  if (slice.length >= 8) {
    const brand = String.fromCharCode(slice[4]!, slice[5]!, slice[6]!, slice[7]!);
    if (brand === "ftyp") return "mp4";
  }
  if (slice[0] === 0x1a && slice[1] === 0x45 && slice[2] === 0xdf) return "mkv";
  try {
    const text = new TextDecoder("utf-8", { fatal: false }).decode(slice.subarray(0, Math.min(slice.length, 24)));
    if (text.trimStart().startsWith("#EXT")) return "hls";
  } catch {
    /* ignore */
  }
  return "unknown";
}

export function mimeForKind(kind: StreamKind, fallback?: string | null): string {
  if (kind === "hls") return "application/vnd.apple.mpegurl";
  if (kind === "ts") return "video/mp2t";
  if (kind === "mp4") return "video/mp4";
  if (kind === "mkv") return "video/x-matroska";
  return fallback || "application/octet-stream";
}

export function looksLikePlaylistUrl(url: string, contentType = ""): boolean {
  return /mpegurl|m3u8/i.test(contentType) || /\.m3u8(\?|$)/i.test(url);
}

export function engineForKind(kind: StreamKind, fallbackUrl?: string): "hls" | "mpegts" | "native" {
  if (kind === "hls") return "hls";
  if (kind === "ts") return "mpegts";
  if (kind === "mp4" || kind === "mkv") return "native";
  if (fallbackUrl && /\.m3u8(\?|$)/i.test(fallbackUrl)) return "hls";
  if (fallbackUrl && /\/live\//i.test(fallbackUrl)) return "mpegts";
  return "native";
}
